File: src/main/java/de/hhu/stups/alloy2b/translation/Alloy2BParser.java
package de.hhu.stups.alloy2b.translation;

import java.io.File;
import java.net.URISyntaxException;
import java.net.URL;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import edu.mit.csail.sdg.alloy4.A4Reporter;
import edu.mit.csail.sdg.alloy4.ConstList;
import edu.mit.csail.sdg.alloy4.Err;
import edu.mit.csail.sdg.alloy4.Pair;
import edu.mit.csail.sdg.alloy4.Pos;
import edu.mit.csail.sdg.alloy4.SafeList;
import edu.mit.csail.sdg.ast.Command;
import edu.mit.csail.sdg.ast.CommandScope;
import edu.mit.csail.sdg.ast.Expr;
import edu.mit.csail.sdg.ast.Func;
import edu.mit.csail.sdg.ast.Sig;
import edu.mit.csail.sdg.parser.CompModule;
import edu.mit.csail.sdg.parser.CompUtil;

import static de.hhu.stups.alloy2b.translation.ParserUtil.sanitizeIdentifier;

/**
 * Convert the abstract syntax tree of an Alloy model to a Prolog term.
 */
public final class Alloy2BParser {
    private final List<Sig> signatures;
    private final List<String> orderedSignatures;
    private final Set<Set<String>> setsOfParents;
    private final List<Sig> enums;
    private final ExpressionToProlog expressionTranslator;

    private ConstList<Command> commands;

    public Alloy2BParser() {
        this.signatures = new ArrayList<>();
        this.orderedSignatures = new ArrayList<>();
        this.setsOfParents = new HashSet<>();
        this.enums = new ArrayList<>();
        this.expressionTranslator = new ExpressionToProlog(this.signatures, this.orderedSignatures, this.setsOfParents);

        this.commands = null; // initialized in translateModule
    }

    private static <T> Stream<T> safeListStream(SafeList<T> safeList) {
        return safeList.makeConstList().stream();
    }

    private String translateModule(CompModule module) {
        this.commands = module.getAllCommands();

        // if the module is named by the user using 'as', module.path is set
        String sModelName = sanitizeIdentifier(module.getModelName());
        String name = module.path.isEmpty()
            ? "(" + sModelName + ",alloy2b_none)"
            : "(" + sModelName + "," + sanitizeIdentifier(module.path) + ")";

        String listOfFacts = safeListStream(module.getAllFacts())
            .map(this::toPrologTerm)
            .collect(Collectors.joining(","));

        String listOfAssertions = module.getAllAssertions().stream()
            .map(this::toPrologTerm)
            .collect(Collectors.joining(","));

        String listOfCommands = this.commands.stream()
            .map(this::toPrologTerm)
            .collect(Collectors.joining(","));

        String listOfFunctions = safeListStream(module.getAllFunc())
            .map(this::toPrologTerm)
            .collect(Collectors.joining(","));

        String listOfSignatures = safeListStream(module.getAllSigs())
            .map(this::toPrologTerm)
            .collect(Collectors.joining(","));

        String parentTypes = "parent_types:" + this.setsOfParents;
        this.setsOfParents.clear();

        return "alloy_model(" + name + ","
            + "facts([" + listOfFacts + "]),"
            + "assertions([" + listOfAssertions + "]),"
            + "commands([" + listOfCommands + "]),"
            + "functions([" + listOfFunctions + "]),"
            + "signatures([" + listOfSignatures + "]),"
            + "ordered_signatures(" + this.orderedSignatures.stream().map(s -> "'" + s + "'").collect(Collectors.toList()) + "),"
            + "[sequences:" + this.expressionTranslator.getUsesSequences() + "," + parentTypes + ","
            + "ordering_successors_only:" + this.expressionTranslator.getOrderingsUseSuccessorsOnly() + "])";
    }

    private static String realPath(String alloyModelPath) {
        try {
            // either from resources
            URL resourceUrl = Alloy2BParser.class.getResource(alloyModelPath);
            if (resourceUrl == null) {
                return alloyModelPath;
            } else {
                try {
                    return new File(resourceUrl.toURI()).toString();
                } catch (URISyntaxException exc) {
                    throw new IllegalArgumentException(exc);
                }
            }
        } catch (IllegalStateException exception) {
            // or an absolute path
            return alloyModelPath;
        }
    }

    private void collectPropertiesFromInclude(CompModule.Open it) {
        if ("util/ordering".equals(it.filename)) {
            List<String> prefixedSignatures = it.args.stream().map(s -> s.replace("this/", "")).collect(Collectors.toList());
            this.orderedSignatures.addAll(prefixedSignatures);
            this.orderedSignatures.add(it.alias);
        }
    }

    public ParserResult parseFromFile(String alloyModelPath) throws Alloy2BParserErr {
        this.orderedSignatures.clear();
        this.expressionTranslator.setUsesSequences(false);
        String path = realPath(alloyModelPath);
        try {
            CompModule astRoot = CompUtil.parseEverything_fromFile(new A4Reporter(), null, path);
            astRoot.getOpens().forEach(this::collectPropertiesFromInclude);
            this.signatures.addAll(astRoot.getRootModule().getAllSigs().makeConstList()); // primitive types are not considered here

            String modules = safeListStream(astRoot.getAllReachableModules()).map(this::translateModule).collect(Collectors.joining(","));
            String rootModule = sanitizeIdentifier(astRoot.getRootModule().getModelName());

            List<String> commandNames = new ArrayList<>();
            for (int i = 0; i < astRoot.getRootModule().getAllCommands().size(); i++) {
                Command cmd = astRoot.getRootModule().getAllCommands().get(i);
                if (cmd.check) {
                    commandNames.add("check" + i);
                } else {
                    commandNames.add("run" + i);
                }
            }

            return new ParserResult("alloy(" + rootModule + ",[" + modules + "]).", ConstList.make(commandNames));
        } catch (Err exception) {
            throw new Alloy2BParserErr(exception);
        }
    }

    private String toPrologTerm(Pair<String, Expr> astNode) {
        return "fact(" + astNode.b.accept(this.expressionTranslator) + ",(" + astNode.b.pos.x + "," + astNode.b.pos.y + "))";
    }

    private String toPrologTerm(Command astNode) {
        String functor = astNode.check ? "check" : "run";

        List<String> exactScopes = astNode.scope.stream()
            .filter(it -> it.isExact)
            .map(Alloy2BParser::createSigScopeTuple)
            .collect(Collectors.toList());

        List<String> upperBoundScopes = astNode.scope.stream()
            .filter(it -> !it.isExact)
            .map(Alloy2BParser::createSigScopeTuple)
            .collect(Collectors.toList());

        return functor + "("
            + astNode.formula.accept(this.expressionTranslator) + ","
            + "global_scope(" + astNode.overall + "),"
            + "exact_scopes(" + exactScopes + "),"
            + "upper_bound_scopes(" + upperBoundScopes + "),"
            + "bitwidth(" + astNode.bitwidth + "),"
            + "maxseq(" + astNode.maxseq + "),"
            + "index(" + this.commands.indexOf(astNode) + "),"
            + "pos(" + astNode.pos.x + "," + astNode.pos.y + "))";
    }

    private static String createSigScopeTuple(CommandScope scope) {
        return "(" + sanitizeIdentifier(scope.sig.label) + "," + scope.startingScope + ")";
    }

    private String toPrologTerm(Func astNode) {
        String functor = astNode.isPred ? "predicate" : "function";
        return functor + "("
            + sanitizeIdentifier(astNode.label) + ","
            + astNode.params().stream().map(this::toPrologTerm).collect(Collectors.toList()) + ","
            + astNode.decls.stream().map(this.expressionTranslator::toPrologTerm).collect(Collectors.toList()) + ","
            + this.toPrologTerm(astNode.getBody()) + ","
            + "pos(" + astNode.pos.x + "," + astNode.pos.y + "))";
    }

    private String toPrologTerm(Expr astNode) {
        return astNode.accept(this.expressionTranslator);
    }

    private String toPrologTerm(Sig astNode) {
        // enums are syntactical sugar: elements are 'one sigs' which extend the base sig
        List<Sig> parentSig = this.enums.stream()
            .filter(it -> astNode.isSameOrDescendentOf(it) && !astNode.isSame(it))
            .collect(Collectors.toList());

        boolean isSubsig = !parentSig.isEmpty();
        String options = isSubsig
            ? "[one,subsig(" + sanitizeIdentifier(((Sig.PrimSig)astNode).parent.label) + ")]"
            : collectSignatureOptionsToPrologList(astNode);

        if (astNode.isEnum != null) {
            this.enums.add(astNode);
        }

        if (!isSubsig) {
            this.signatures.add(astNode);
        }

        return "signature("
            + sanitizeIdentifier(astNode.label) + ","
            + "[" + safeListStream(astNode.getFieldDecls()).map(this.expressionTranslator::toPrologTerm).collect(Collectors.joining(",")) + "],"
            + "[" + safeListStream(astNode.getFacts()).map(this::toPrologTerm).collect(Collectors.joining(",")) + "],"
            + options + ","
            + "pos(" + astNode.pos.x + "," + astNode.pos.y + "))";
    }

    private String collectSignatureOptionsToPrologList(Sig astNode) {
        List<String> lstOptions = new ArrayList<>();
        String sigName = astNode.label.replace("this/", "");
        if (astNode.isEnum != null) {
            // enums are treated as orderings in Alloy
            lstOptions.add("enum");
            this.orderedSignatures.remove(sigName);
        }
        if (this.orderedSignatures.contains(sigName)) {
            lstOptions.add("ordered");
        }
        if (astNode.isAbstract != null) {
            lstOptions.add("abstract");
        }
        if (astNode.isLone != null) {
            lstOptions.add("lone");
        }
        if (astNode.isMeta != null) {
            lstOptions.add("meta");
        }
        if (astNode.isOne != null) {
            lstOptions.add("one");
        }
        if (astNode.isPrivate != null) {
            lstOptions.add("private");
        }
        if (astNode.isSome != null) {
            lstOptions.add("some");
        }
        if (astNode.isSubset != null) {
            lstOptions.add("subset(" + ((Sig.SubsetSig)astNode).parents.stream().map(it -> sanitizeIdentifier(it.label)).collect(Collectors.toList()) + ")");
        }
        if (isExtendingSignature(astNode)) {
            lstOptions.add("subsig(" + sanitizeIdentifier(((Sig.PrimSig)astNode).parent.label) + ")");
        }
        return lstOptions.toString();
    }

    private boolean isExtendingSignature(Sig sig) {
        Pos isSubSig = sig.isSubsig;
        return sig instanceof Sig.PrimSig && isSubSig != null && (isSubSig.x != isSubSig.x2 || isSubSig.y != isSubSig.y2);
    }
}

File: src/main/kotlin/de/hhu/stups/alloy2b/translation/Alloy2BParser.kt
package de.hhu.stups.alloy2b.translation

import edu.mit.csail.sdg.alloy4.A4Reporter
import edu.mit.csail.sdg.alloy4.ConstList
import edu.mit.csail.sdg.alloy4.Err
import edu.mit.csail.sdg.alloy4.Pair
import edu.mit.csail.sdg.ast.*
import edu.mit.csail.sdg.parser.CompModule
import edu.mit.csail.sdg.parser.CompUtil

data class ParserResult(val prologTerm: String, val commandNames: ConstList<String>)

data class Alloy2BParserErr(val msg: String, val filename: String,
                            val colStart: Int, val rowStart: Int, val colEnd: Int, val rowEnd: Int) : Exception(msg)

/**
 * Convert the abstract syntax tree of an Alloy model to a Prolog term.
 */
class Alloy2BParser {

    private val signatures = mutableListOf<Sig>()
    private val orderedSignatures = mutableListOf<String>()
    private val setsOfParents = mutableSetOf<MutableSet<String>>()
    private val enums = mutableListOf<Sig>()
    private val expressionTranslator = ExpressionToProlog(signatures, orderedSignatures, setsOfParents)

    private lateinit var commands: ConstList<Command>

    private fun translateModule(module: CompModule): String {
        commands = module.allCommands
        // if the module is named by the user using 'as', module.path is set
        val sModelName = sanitizeIdentifier(module.modelName)
        val name = if (module.path.isEmpty()) "($sModelName,alloy2b_none)" else
            "($sModelName,${sanitizeIdentifier(module.path)})"
        val listOfFacts = module.allFacts.joinToString(",") { toPrologTerm(it) }
        val listOfAssertions = module.allAssertions.joinToString(",") { toPrologTerm(it) }
        val listOfCommands = commands.joinToString(",") { toPrologTerm(it) }
        val listOfFunctions = module.allFunc.joinToString(",") { toPrologTerm(it) }
        val listOfSignatures = module.allSigs.joinToString(",") { toPrologTerm(it) }
        val parentTypes = "parent_types:$setsOfParents"
        setsOfParents.clear()
        return "alloy_model($name,facts([$listOfFacts]),assertions([$listOfAssertions]),commands([$listOfCommands])," +
                "functions([$listOfFunctions]),signatures([$listOfSignatures])," +
                "ordered_signatures(${orderedSignatures.map { "'$it'" }})," +
                "[sequences:${expressionTranslator.usesSequences},$parentTypes," +
                "ordering_successors_only:${expressionTranslator.orderingsUseSuccessorsOnly}])"
    }

    private fun realPath(alloyModelPath: String): String {
        return try {
            // either from resources
            object {}.javaClass.getResource(alloyModelPath).file
        } catch (exception: IllegalStateException) {
            // or an absolute path
            alloyModelPath
        }
    }

    private fun collectPropertiesFromInclude(it: CompModule.Open) {
        if ("util/ordering" == it.filename) {
            val prefixedSignatures = it.args.map { it.replace("this/", "") }
            orderedSignatures.addAll(prefixedSignatures)
            orderedSignatures.add(it.alias)
        }
    }

    @Throws(Alloy2BParserErr::class)
    fun parseFromFile(alloyModelPath: String): ParserResult {
        orderedSignatures.clear()
        expressionTranslator.usesSequences = false
        val path = realPath(alloyModelPath)
        try {
            val astRoot = CompUtil.parseEverything_fromFile(A4Reporter(), null, path)
            astRoot.opens.forEach { collectPropertiesFromInclude(it) }
            signatures.addAll(astRoot.rootModule.allSigs) // primitive types are not considered here

            val modules = astRoot.allReachableModules.joinToString(",", transform = ::translateModule)
            val rootModule = sanitizeIdentifier(astRoot.rootModule.modelName)
            val commandNames: ConstList<String> = ConstList.make(astRoot.rootModule.allCommands
                    .mapIndexed { i, cmd -> if (cmd.check) "check$i" else "run$i" })
            return ParserResult("alloy($rootModule,[$modules]).", commandNames)
        } catch (exception: Err) {
            val pos = exception.pos
            throw Alloy2BParserErr(exception.msg, exception.pos.filename, pos.x, pos.y, pos.x2, pos.y2)
        }
    }

    private fun toPrologTerm(astNode: Pair<String, Expr>) =
            "fact(${astNode.b.accept(expressionTranslator)},(${astNode.b?.pos?.x},${astNode.b?.pos?.y}))"

    private fun toPrologTerm(astNode: Command): String {
        val functor = if (astNode.check) "check" else "run"
        val exactScopes = astNode.scope.asSequence().filter { it.isExact }
                .map { createSigScopeTuple(it) }.toList()
        val upperBoundScopes = astNode.scope.asSequence().filter { !it.isExact }
                .map { createSigScopeTuple(it) }.toList()
        return "$functor(${astNode.formula.accept(expressionTranslator)},global_scope(${astNode.overall})," +
                "exact_scopes($exactScopes)," +
                "upper_bound_scopes($upperBoundScopes)," +
                "bitwidth(${astNode.bitwidth}),maxseq(${astNode.maxseq}),index(${commands.indexOf(astNode)})," +
                "pos(${astNode.pos.x},${astNode.pos.y}))"
    }

    private fun createSigScopeTuple(scope: CommandScope) =
            "(${sanitizeIdentifier(scope.sig.label)},${scope.startingScope})"

    private fun toPrologTerm(astNode: Func): String {
        val functor = if (astNode.isPred) "predicate" else "function"
        return "$functor(${sanitizeIdentifier(astNode.label)},${astNode.params().map { toPrologTerm(it) }}," +
                astNode.decls.map { expressionTranslator.toPrologTerm(it) } +
                ",${toPrologTerm(astNode.body)},pos(${astNode.pos.x},${astNode.pos.y}))"
    }

    private fun toPrologTerm(astNode: Expr) =
            astNode.accept(expressionTranslator)

    private fun toPrologTerm(astNode: Sig): String {
        // enums are syntactical sugar: elements are 'one sigs' which extend the base sig
        val parentSig = enums.filter { astNode.isSameOrDescendentOf(it) && !astNode.isSame(it) }
        val isSubsig = parentSig.isNotEmpty()
        val options =
                if (isSubsig) "[one,subsig(${sanitizeIdentifier((astNode as Sig.PrimSig).parent.label)})]"
                else collectSignatureOptionsToPrologList(astNode)
        if (astNode.isEnum != null) {
            enums.add(astNode)
        }
        if (!isSubsig) {
            signatures.add(astNode)
        }
        return "signature(${sanitizeIdentifier(astNode.label)}," +
                "[${astNode.fieldDecls.joinToString(",") { expressionTranslator.toPrologTerm(it) }}]," +
                "[${astNode.facts.joinToString(",") { toPrologTerm(it) }}]," +
                "$options,pos(${astNode.pos.x},${astNode.pos.y}))"
    }

    private fun collectSignatureOptionsToPrologList(astNode: Sig): String {
        val lstOptions = mutableListOf<String>()
        val sigName = astNode.label.replace("this/", "")
        if (astNode.isEnum != null) {
            // enums are treated as orderings in Alloy
            lstOptions.add("enum")
            orderedSignatures.remove(sigName)
        }
        if (orderedSignatures.contains(sigName)) {
            lstOptions.add("ordered")
        }
        if (astNode.isAbstract != null) {
            lstOptions.add("abstract")
        }
        if (astNode.isLone != null) {
            lstOptions.add("lone")
        }
        if (astNode.isMeta != null) {
            lstOptions.add("meta")
        }
        if (astNode.isOne != null) {
            lstOptions.add("one")
        }
        if (astNode.isPrivate != null) {
            lstOptions.add("private")
        }
        if (astNode.isSome != null) {
            lstOptions.add("some")
        }
        if (astNode.isSubset != null) {
            lstOptions.add("subset(${(astNode as Sig.SubsetSig).parents.map { sanitizeIdentifier(it.label) }})")
        }
        if (isExtendingSignature(astNode)) {
            lstOptions.add("subsig(${sanitizeIdentifier((astNode as Sig.PrimSig).parent.label)})")
        }
        return lstOptions.toString()
    }

    private fun isExtendingSignature(sig: Sig): Boolean {
        val isSubSig = sig.isSubsig
        return sig is Sig.PrimSig && isSubSig != null && (isSubSig.x != isSubSig.x2 || isSubSig.y != isSubSig.y2)
    }
}
